### PR TITLE
Use parking_lot::RwLock for PropertyDeclarationBlock

### PR DESCRIPTION
**Pull request by SimonSapin**
As discussed in https://bugzilla.mozilla.org/show_bug.cgi?id=1305141

Closes #13176

---

Original PR title: Stop relying on `impl<T: HeapSizeOf> HeapSizeOf for Arc<T>`
https://github.com/servo/heapsize/issues/37#issuecomment-249861171

This builds on top of that.

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

- [ ] There are tests for these changes OR
- [x] These changes do not require tests because refactor

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/13459)

### CODE REVIEW

**Comment by nox** [comment]
Reviewed 3 of 3 files at r1, 12 of 12 files at r2, 8 of 8 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/13459)*

File: components/script/dom/cssstyledeclaration.rs [hunk 1]
```diff
@@ -119,43 +118,42 @@ impl CSSStyleDeclarationMethods for CSSStyleDeclaration {
 
         // Step 2
         if let Some(shorthand) = Shorthand::from_name(&property) {
+            let style_attribute = owner.style_attribute().borrow();
+            let style_attribute = if let Some(ref style_attribute) = *style_attribute {
```
Comment (emilio) [inline components/script/dom/cssstyledeclaration.rs]: This might be easier to read with a `match` instead?

**Review by emilio — APPROVED** [review]
I'm fine with this. In general, I'd like us to ensure we don't block in most cases (it should only be needed for keyframes), but I see why that's hard. It can become a followup. In any case it's better than the YOLO RefCell we have now.

**Comment by SimonSapin** [comment]
@bors-servo r=emilio p=10

(Prioritized as this is prone to bitrot.)

**Comment by SimonSapin** [comment]
@bors-servo r=emilio

Fixed order of `use` statements.

**Comment by SimonSapin** [comment]
Uh oh.

```
  ▶ FAIL [expected PASS] /_mozilla/css/iframe/hide_after_load.html
  └   → /_mozilla/css/iframe/hide_after_load.html 7e556c519a46ddc020581d2275b67ff4c39cd5d8
/_mozilla/css/iframe/hide_after_load_ref.html ce129a418aad4e23134dc79329a8508de6be8223
Testing 7e556c519a46ddc020581d2275b67ff4c39cd5d8 == ce129a418aad4e23134dc79329a8508de6be8223

  ▶ Unexpected subtest result in /_mozilla/mozilla/calc.html:
  │ FAIL [expected PASS] calc(10px - 5px)
  │   → assert_equals: expected "5px" but got "20px"
  │ FAIL [expected PASS] calc(10px / 2)
  │   → assert_equals: expected "5px" but got "20px"
  │ FAIL [expected PASS] calc(10px + 10em - 10px)
  │   → assert_equals: expected "160px" but got "10px"
  │ FAIL [expected PASS] calc(1px + 1pt + 1pc + 1in + 1cm + 1mm)
  │   → assert_equals: expected "155.88333333333333px" but got "20px"
  │ FAIL [expected PASS] calc(((3 - 1) * (8 + 4)) * 10px)
  │   → assert_equals: expected "240px" but got "170px"
  │ FAIL [expected PASS] calc(5 * (20px / 2 + 7 * (3em + 12px/4 + (8 - 2) * 2rem)))
  │   → assert_equals: expected "8555px" but got "10px"
  │ 
  │ @http://web-platform.test:8000/_mozilla/mozilla/calc.html:45:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1401:20
  │ test@http://web-platform.test:8000/resources/testharness.js:501:9
  │ @http://web-platform.test:8000/_mozilla/mozilla/calc.html:42:5
  └ @http://web-platform.test:8000/_mozilla/mozilla/calc.html:41:1

  ▶ Unexpected subtest result in /css-backgrounds-3_dev/html4/background-clip-001.htm:
  │ FAIL [expected PASS] background-clip_content-box
  │   → assert_equals: background-clip supporting value expected "content-box" but got "border-box"
  │ 
  │ @http://web-platform.test:8000/css-backgrounds-3_dev/html4/background-clip-001.htm:21:13
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1227:20
  │ test@http://web-platform.test:8000/resources/testharness.js:428:9
  └ @http://web-platform.test:8000/css-backgrounds-3_dev/html4/background-clip-001.htm:20:9

  ▶ Unexpected subtest result in /css-backgrounds-3_dev/html4/background-clip-001.htm:
  │ FAIL [expected PASS] background-clip_padding-box
  │   → assert_equals: background-clip supporting value expected "padding-box" but got "border-box"
  │ 
  │ @http://web-platform.test:8000/css-backgrounds-3_dev/html4/background-clip-001.htm:15:13
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1227:20
  │ test@http://web-platform.test:8000/resources/testharness.js:428:9
  └ @http://web-platform.test:8000/css-backgrounds-3_dev/html4/background-clip-001.htm:14:9

  ▶ Unexpected subtest result in /css-backgrounds-3_dev/html4/background-origin-001.htm:
  │ FAIL [expected PASS] background-origin_padding-box
  │   → assert_equals: background-origin supporting value expected "padding-box" but got "border-box"
  │ 
  │ @http://web-platform.test:8000/css-backgrounds-3_dev/html4/background-origin-001.htm:15:13
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1227:20
  │ test@http://web-platform.test:8000/resources/testharness.js:428:9
  └ @http://web-platform.test:8000/css-backgrounds-3_dev/html4/background-origin-001.htm:14:9

  ▶ Unexpected subtest result in /css-backgrounds-3_dev/html4/background-size-001.htm:
  │ FAIL [expected PASS] background-size_auto_auto
  │   → assert_equals: background-size supporting value expected "auto auto" but got "contain"
  │ 
  │ @http://web-platform.test:8000/css-backgrounds-3_dev/html4/background-size-001.htm:69:13
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1227:20
  │ test@http://web-platform.test:8000/resources/testharness.js:428:9
  └ @http://web-platform.test:8000/css-backgrounds-3_dev/html4/background-size-001.htm:68:9

  ▶ Unexpected subtest result in /css-backgrounds-3_dev/html4/background-size-001.htm:
  │ FAIL [expected PASS] background-size_length_percentage
  │   → assert_equals: background-size supporting value expected "15px 50%" but got "0px auto"
  │ 
  │ @http://web-platform.test:8000/css-backgrounds-3_dev/html4/background-size-001.htm:99:13
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1227:20
  │ test@http://web-platform.test:8000/resources/testharness.js:428:9
  └ @http://web-platform.test:8000/css-backgrounds-3_dev/html4/background-size-001.htm:98:9

  ▶ Unexpected subtest result in /css-backgrounds-3_dev/html4/background-size-001.htm:
  │ FAIL [expected PASS] background-size_percentage_auto
  │   → assert_equals: background-size supporting value expected "50% auto" but got "0px auto"
  │ 
  │ @http://web-platform.test:8000/css-backgrounds-3_dev/html4/background-size-001.htm:105:13
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1227:20
  │ test@http://web-platform.test:8000/resources/testharness.js:428:9
  └ @http://web-platform.test:8000/css-backgrounds-3_dev/html4/background-size-001.htm:104:9

  ▶ Unexpected subtest result in /css-backgrounds-3_dev/html4/background-size-001.htm:
  │ FAIL [expected PASS] background-size_auto_percentage
  │   → assert_equals: background-size supporting value expected "auto 50%" but got "0px auto"
  │ 
  │ @http://web-platform.test:8000/css-backgrounds-3_dev/html4/background-size-001.htm:81:13
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1227:20
  │ test@http://web-platform.test:8000/resources/testharness.js:428:9
  └ @http://web-platform.test:8000/css-backgrounds-3_dev/html4/background-size-001.htm:80:9
```

I don’t know what’s going on here, but there’s no known intermittent filed for these. Will need to investigate.

**Comment by SimonSapin** [comment]
The only (I think) behavior change is that this PR removes some `Arc::make_mut` and uses locking instead. So a `PropertyDeclarationBlock` that could be referenced elsewhere and was previously immutable is now mutable. This seemed like a good idea, reducing the amount of allocation and cloning.

Disabling the applicable declaration cache (by making `PartialEq for ApplicableDeclarationsCacheQuery` always return false) seems to fix the issue. This is likely because the cache uses the address of a `Arc<PropertyDeclarationBlock>` (now `Arc<RwLock<PropertyDeclarationBlock>>`) pointer as its hash / identity.

I can’t think of proper fix that doesn’t revert to the previous copy-on-write behavior of using `Arc::make_mut` (or equivalent) in CSSOM (including `Element.style`), cloning blocks when modifying them if they have more than one reference.

@pcwalton, any idea?

**Comment by emilio** [comment]
@SimonSapin: Flushing the applicable declarations cache after a restyle should be fine, right?

**Comment by SimonSapin** [comment]
Interesting. Or maybe have any `PropertyDeclarationBlock` mutation mark the cache as dirty? That way we can still keep it between restyles for DOM updates.

**Comment by emilio** [comment]
Por even better, maybe just avoid adding to the applicable declaration cache if the matched element has a style attribute?

Seems like it's impossible for that to match in the same restyle with any other element, and I don't think we care a lot about the case where there are such a little amount of elements that the entry persists in the cache across restyles, so it might even be a win?

**Comment by SimonSapin** [comment]
We’ll still need some way to deal with this cache when style rules are mutated, when we eventually support CSSOM.

**Comment by SimonSapin** [comment]
Rebased and added a "Disable applicable declarations cache for elements with a style attribute" commit.

We’ll need to do something (maybe mark the cache as dirty whenever a `PropertyDeclarationBlock` is mutated?) when more of CSSOM is implemented, but in the meantime I’d like to land this before it bit-rots again.

re-r? @emilio

**Comment by emilio** [comment]
Also, I think this is the only not-previously-reviewed commit, so feel free to r=me the whole PR when that comment is added.

@bors-servo: try

**Comment by SimonSapin** [comment]
I, that was the only new significant change.

@bors-servo r=emilio

**Pull request merged**